Tutorial on creating a bare metal application (#1305)
ARMmbed/Handbook: docs/program-setup/bare_metal/baremetal_usage.md

> # Using bare metal profile

Reviewer: ```suggestion
# Using the bare metal profile
```

> # Using bare metal profile

Reviewer: This requires an intro that explains why you need a specific workflow, and what this page covers.

Reviewer: ```suggestion
# Using bare metal profile

This guide shows how to create a bare metal profile application: 

1. By default, when you build an application binary, the build tool uses the full profile. To use the bare metal profile, you need to set up your application to override this default behaviour. 
1. Bare metal has a minimal set of APIs. You can add additional ones [from the list of supported API]().
1. Bare metal has its own set of tests in Greentea that can verify <!--what does it verify? just the default APIs?-->
```

Author: > what does it verify? just the default APIs?

It's a really good question. _As far as I understand_, we initially wanted to verify the default APIs only. But some of the tests couldn't compile, so we brought in more APIs just to fix the existing compilation issue. The side effect? More APIs are tested! (https://github.com/ARMmbed/mbed-os/pull/11721)

~With this explanation, I don't think there is a clear logic why we verify some APIs and not others with the bare metal profile, but that's just my personally understanding.~ I'm wondering if all APIs that don't require RTOS get covered as a result, in which case it'll be easier to explain to users. I will discuss this with the team.

Author: Okay, it's simply all tests that currently support the bare metal profile (and the target, obviously)

> ## Creating a bare metal application

Reviewer: I will change all of this to numbered steps. This is easier for users when they have to go do something and then come back to a document.

> First, clone Mbed OS into the directory where we want to create an application

Reviewer: ```suggestion
1. Clone Mbed OS into the directory where you want to create an application:
```

> First, clone Mbed OS into the directory where we want to create an application

Reviewer 2: I would like this tutorial to be a simple example with events. See https://github.com/ARMmbed/mbed-os-example-baremetal-eventqueue-blinky

> ```

Reviewer: I'll handle the indents later to support a numbered list

mkdir my_mbed_app && cd my_mbed_app
git clone https://github.com/ARMmbed/mbed-os.git
```

> To enable Mbed OS bare metal, create an `mbed_app.json` with the following contents:

Reviewer: ```suggestion
1. To use the bare metal profile, create an `mbed_app.json` with the following contents in your application's root directory:
```

```
{

>     "requires": ["bare-metal"]

Reviewer: Should show small C library use and link to the full explanation

https://os.mbed.com/docs/mbed-os/bare-metal-work/bare-metal/using-small-c-libraries.html
(can see that page on the VPN)

Author: That's a good point

}
```

> On bare metal, features are enabled with a bottom-up approach - those listed in [Bare metal APIs](bare_metal_api.md#bare-metal-apis) are now ready for use (but not other features) in our application. See [Bare metal example](baremetal_example.md).

Reviewer: ```suggestion
Bare metal has a minimal set of default APIs - those that are always available to a bare metal application. You can add other supported APIs if you need the features they enable. 

For a list of default and supported APIs, [please see our full API list](../apis/index.html).
```

> To pull in additional features (take filesystem for example)

Reviewer: ```suggestion
To add API - in this example, the FileSystem class:
```

> To pull in additional features (take filesystem for example)

Reviewer 2: As discussed, FileSystem is quite complex for simple IOT devices. Let's use an event queue instead.

> 1. Locate the libraries we want to use in `mbed-os/` and find their names in their respective `mbed_lib.json`, e.g. `features/storage/filesystem/mbed_lib.json`:

Reviewer: ```suggestion
1. Locate the library you want to use in `mbed-os/`.
1. In the library folder, find the library's name in `mbed_lib.json`. You will need it for the next step.

     For example: `features/storage/filesystem/mbed_lib.json`:
```

Reviewer: I have to say I'm not sure about using "library" here, because in some contexts it has a more specific meaning. And what we're really looking for here is either .cpp or .h. 

I think this needs clarifying.

    ```json
    {
        "name": "filesystem",
        "config": {
            "present": 1
        }
    }
    ```

> 1. Add the libraries to `"requires"` in our application's `mbed_app.json`, for example:

Reviewer: ```suggestion
1. Find the file `mbed_app.json` file in the application's root directory, and add the library to the `"requires"` field:
```

    ```json
    {
        "requires": [
            "bare-metal",
            "filesystem"
        ]
    }
    ```

> 1. Write the application to use bare metal APIs and libraries added in 2.

Reviewer: ```suggestion
1. Write the application using the libraries you added.
```

> 1. Try to compile:

Reviewer: ```suggestion
1. Compile your application:
```

    ```
    mbed compile -t <TOOLCHAIN> -m <TARGET>
    ```

>     Compilation may fail if any libaries added in 2 depend on other libraries. Use error messages to locate dependent libraries (like we did in 1.) and add them to `mbed_lib.json` too, for example `"filesystem"` depends on `"fat_chan"`:

Reviewer: ```suggestion
    If the library you added depends on additional libraries, the compilation may fail. Use the build's error messages to locate dependencies and add them to `mbed_lib.json`, as you did for FileSystem. 
    
    For example `"filesystem"` depends on `"fat_chan"`:
```

    ```
    {
        "requires": [
            "bare-metal",
            "filesystem",
            "fat_chan"
        ]
    }
    ```

>     Repeat this step until compilation succeeds, i.e. we have added all dependencies.

Reviewer: ```suggestion
    Repeat this step until compilation succeeds (that is, until you have added all dependencies).
```

> ## Running Greentea tests for bare metal profile

Reviewer: ```suggestion
## Running Greentea tests for bare metal applications
```

(If you have not used Greentea before, [you can learn more in our documentation](../tools/greentea-testing-applications.html).)

> To run Greentea tests for bare metal profile, simply use `baremetal.json`. The full commands are as follows:

Reviewer: ```suggestion
To run Greentea tests for a bare metal application, use Greentea's `baremetal.json`:
```

> 1. Change directory:

Reviewer: Where was I to begin with? This is my application's mbed-os directory, right?

    ```
    cd mbed-os
    ```

1. Execute the Greentea test suite with the bare metal configuration for the supported toolchains:

    ```

>     mbed test -m <YOUR_TARGET> -t <TOOLCHAIN> --app-config TESTS/configs/baremetal.json

Reviewer 2: Should we use `--app-config` or `--test-config`?
Can this command also be used with a modified mbed_app.json to change library defaults?

Author: Yea `--test-config` is more sensible.

From the test command parser:
https://github.com/ARMmbed/mbed-os/blob/af4c8a94f3913ceda9069ec79cba81a793d695fe/tools/test.py#L177-L189
Unfortunately (as per the comment there) if `--test-config` (or `--app-config`) is used, the test system won't look into the app-provided `mbed_app.json`.

Reviewer 2: Let's keep app-config so we don't confuse users with tests configs.
